### PR TITLE
Add option to use bf16 in PT sdp (#5)

### PR DESCRIPTION
**Pull request by astachowiczhabana**
The new option --sdp_on_bf16 allows pyTorch to use reduced precision in sdp in the math backend.
The change affects the following examples:
* contrastive-image-text/run_bridgetower.py
* question-answering/run_qa.py
* speech-recognition/run_speech_recognition_ctc.py
* stable-diffusion//text_to_image_generation.py
* text-classification/run_glue.py

### CODE REVIEW

**Comment by jiminha** [comment]
@astachowiczhabana  which model and configuration are you running with this flag enabled? We'd like to do the same for the OH CI/PYTEST.

**Comment by hsubramony** [comment]
@astachowiczhabana  dont we need to update README for all the models with this option ?

**Comment by ugolowic** [comment]
@jiminha

Inference - 1 card:
- question-answering/run_qa.py bert-large-uncased-whole-word-masking bs 8
- text-classification/run_glue.py bert-large-uncased-whole-word-masking bs 8
- stable-diffusion/text_to_image_generation.py stabilityai/stable-diffusion-2-base bs 4
- speech-recognition/run_speech_recognition_ctc.pywav2vec2-librispeech-clean-100h-demo-dist bs 128
- contrastive-image-text/run_bridgetower.py BridgeTower/bridgetower-large-itm-mlm-itc bs 16

Training - 8 cards:
- audio-classification/run_audio_classification.py facebook/wav2vec2-base bs 16
- image-classification/run_image_classification.py google/vit-base-patch16-224-in21k bs 64

File: examples/audio-classification/run_audio_classification.py [hunk 1]
```diff
@@ -220,6 +221,9 @@ def main():
         token=model_args.token,
     )
 
+    if training_args.sdp_on_bf16:
+        torch._C._set_math_sdp_allow_fp16_bf16_reduction(True)
```
Comment (regisss) [inline examples/audio-classification/run_audio_classification.py:225]: It would be easier to do it in the training args post_init (or setup_devices) method I think, WDYT? https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/training_args.py#L832
Same for diffusion pipelines, we could set this directly there.

Comment (ugolowic) [reply examples/audio-classification/run_audio_classification.py:225]: The diffusion example doesn't use GaudiTrainingArgs. Where do you think the bf16 reduction should be called in this case?

Comment (regisss) [reply examples/audio-classification/run_audio_classification.py:225]: Sorry it was a bit unclear, I was referring to the pipelines themselves. I think this should go in the `__init__` of the `GaudiDiffusionPipeline` class: https://github.com/huggingface/optimum-habana/blob/48423761ff3039095f9b4408b73c9e094b302d5c/optimum/habana/diffusers/pipelines/pipeline_utils.py#L118

Comment (ugolowic) [reply examples/audio-classification/run_audio_classification.py]: I've made the changes. While with training args the code is simpler, I'm not so sure with the pipelines. Tell me what you think.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -148,6 +150,7 @@ def __init__(
         use_hpu_graphs: bool = False,
         gaudi_config: Union[str, GaudiConfig] = None,
         bf16_full_eval: bool = False,
+        sdp_on_bf16: bool = False,
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:153]: We need to add this arg to the call at line 155 too, such as:
```py
GaudiDiffusionPipeline.__init__(
    self,
    use_habana,
    use_hpu_graphs,
    gaudi_config,
    bf16_full_eval,
    sdp_on_bf16,
 )
```

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -131,6 +131,8 @@ class GaudiStableDiffusionPipeline(GaudiDiffusionPipeline, StableDiffusionPipeli
         bf16_full_eval (bool, defaults to `False`):
             Whether to use full bfloat16 evaluation instead of 32-bit.
             This will be faster and save memory compared to fp32/mixed precision but can harm generated images.
+        sdp_on_bf16 (bool, defaults to `False`):
+            Whether to allow pyTorch to use reduced precision in the SDPA math backend.
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:135]:
```suggestion
            Whether to allow PyTorch to use reduced precision in the SDPA math backend.
```

File: optimum/habana/diffusers/pipelines/pipeline_utils.py [hunk 4]
```diff
@@ -112,6 +112,8 @@ class GaudiDiffusionPipeline(DiffusionPipeline):
         bf16_full_eval (bool, defaults to `False`):
             Whether to use full bfloat16 evaluation instead of 32-bit.
             This will be faster and save memory compared to fp32/mixed precision but can harm generated images.
+        sdp_on_bf16 (bool, defaults to `False`):
+            Whether to allow pyTorch to use reduced precision in the SDPA math backend.
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/pipeline_utils.py:116]:
```suggestion
            Whether to allow PyTorch to use reduced precision in the SDPA math backend.
```

File: optimum/habana/transformers/training_args.py [hunk 5]
```diff
@@ -300,6 +300,20 @@ class GaudiTrainingArguments(TrainingArguments):
...
     )
 
+    # Use this to override default attn_implementation in transformers
+    attn_implementation: Optional[str] = field(
+        default="eager",
+        metadata={
+            "help": "choose whether to use scale dot product attention (SDPA) or not.",
+            "choices": ["eager", "sdpa"],
+        },
+    )
```
Comment (regisss) [inline optimum/habana/transformers/training_args.py:310]: Where is this used?

**Review by regisss — COMMENTED** [review]
Thanks for making the suggested changes! We need to add the arg in the pipeline init call too (see my comment below) and then I'll merge this PR :slightly_smiling_face: 

I think it's better to do it that way for the diffusion pipelines because users simply have one arg to give to enable it VS adding themselves the call to `torch._C._set_math_sdp_allow_fp16_bf16_reduction`.

**Review by regisss — APPROVED** [review]
